RFC.md

```diff
@@ -1,12 +1,155 @@
+void set_subarray(unsigned n0, unsigned n1,
+	unsigned o0, unsigned o1,
+	unsigned s0, unsigned s1,
+	float A[n0][n1]) {
+	for (unsigned i = 0; i < s0; i++)
+		for (unsigned j = 0; j < s1; j++)
+				S: A[i + o0][j + o1] = 1;
+}
+```
+We first reduce this index expression to a sum of products:
+
+```
+(i + o0) * n0 + (j + o1) = n0i + n0o0 + j + o1
+ix(i, j, n0, n1, o0, o1) = n0i + n0o0 + j + o1
+```
+
+`ix` is the index expression which `LLVM` will see, since it is fully
+flattened, in comparison with the multi-dimensional index expression
+`index:[i + o0][j + o1] | sizes:[][n1]`.
+
+We will now show _why_ this multi-dimensional index is not always correct,
+and why guessing for one language will not work for another:
+
+Consider a call `set_subarray(n0=8, n1=9, o0=4, o1=6, s0=3, s1=6)`. At face
+value, this is incorrect if we view the array as 2D grid, since the size
+of the array is `(n0, n1) = (8, 9)`, but we are writing an array of size
+`(s0, s1) = (3, 6)` starting from `(o0, o1) = (4, 6)`. Clearly, we will
+exceed the width of the array, since `(s1 + o1 = 6 + 6 = 12) > (n1 = 9)`.
+However, now think of the array as a flattened 1D representation. In this
+case, the total size of the array is `n1xn2 = 8x9 = 72`, while the largest
+element we will access is at the largest value of `(i, j)`. That is, 
+`i = s0 - 1 = 2`, and `j = s1 - 1 = 5`.
+
+The largest index will be `ix(i=2, j=5, n0=8, n1=9, o0=4, o1=6) = 8*2+8*4+5+6=59`.
+Since `59 < 72`, we are clearly at _legal_ array indeces, by C semantics!
+
+The definition of the semantics of the language **changed the illegal
+multidimensional access** (which is illegal since it exceeds the `n1`
+dimension), into a **legal flattened 1D access** (which is legal since the
+flattened array indeces are inbounds).
+
+LLVM has no way of expressing these two different semantics. Hence, we are
+forced to:
+1. Consider flattened 1D accesses, which makes analysis of index expressions
+equivalent to analysis of polynomials over the integers, which is hard.
+2. Guess multidimensional representations, and use them at the express of
+soundness bugs as shown above.
+3. Guess multidimensional representations, use them, and check their validity
+at runtime, causing a runtime performance hit.
+
+
+Currently, Polly opts for option (3), which is to emit runtime checks, which
+will cause polly to bail out on the above program on the shown parameter
+configuration, since it detects that the multidimensional view that was
+guessed is incorrect.
+
+Ideally, we would like a mechanism to directly express the multidimensional
+semantics, which would eliminate this kind of guesswork from Polly/LLVM,
+which would both make code faster, and easier to analyze.
+
+
+## The request for change: A new instruction, `multidim_array_index`.
+
+We propose the addition of a new instruction, called `multidim_array_index`,
+which allows one to represent such multidimensional array accesses without
+flattening.
```

Comment:
I think it's important to sketch a proposal including specifics here. We have the name `multidim_array_index`, which is specific. What are the arguments to it?

When we post the RFC to the LLVM list, I think we should try to take it in two steps:
 1. Reach agreement on the general direction (e.g. adding some sort of intrinsic for this problem)
 2. Reach agreement on the exact details (e.g. the intrinsic name and the arguments and what they mean).

I think the 1st RFC we send should present a concrete answer to (2) while also saying that it is a starting point for discussion and that we are expecting to adjust the exact proposal based upon feedback.